_episodes/l1-05-test_doubles.md

```diff
@@ -0,0 +1,239 @@
+
+[Wikipedia]: https://en.wikipedia.org/wiki/Test_double
+
+### Dependency injection
+
+Consider the following function that normalizes an array according to some definition of norm (ignore whether this is the most performing approach or not):
+
+```cpp
+void normalize_v1(int array[], int length)
+{
+    double norm{calculate_norm(array)};
+
+    for (int i{0}; i < length; ++i)
+    {
+        array[i] /= norm;
+    }
+}
+```
+
+You need to test it, but you do not want to have to calculate the norm along the way. How would you tell the `normalize` function to use a test double for `calculate_norm` and not the real one?
+
+Well, you cannot. `calculate_norm` is hardcoded in the definition of `normalize` so replacing it for another function is not possible, especially if we consider that `normalize` will be defined in a particular file within your code, but you are testing it somewhere else.
+
+Now consider the following alternative version of the function `normalize`:
+
+```cpp
+void normalize_v2(
+    int array[],
+    int length,
+    std::function<double(int[])> func = calculate_norm
+    )
+{
+    double norm{func(array)};
+
+    for (int i{0}; i < length; ++i)
+    {
+        array[i] /= norm;
+    }
+}
+```
+
+Compared with the first version, this `normalize` function does exactly the same thing and can be invoked in exactly the same way but, in addition, you can optionally control what specific function is used to calculate the norm. In particular, you can provide a test double that replaces the default `calculate_norm`.
+
+This is called [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection) and its application goes well beyond testing: it helps make the code more modular and re-usable by making it less intrinsically linked to specific design choices or dependencies. In the example, we could use a different definition of norm - [and there are quite a few!](https://en.wikipedia.org/wiki/Norm_(mathematics))
+
+> ### Dependency injection is an important design pattern
+>
+> Do not disregard the value of dependency injection as an approach only useful in testing. If you design your code with dependency injection in mind, it will become more flexible and powerful. Here we have presented just one way of doing dependency injection, but there are other approaches that might be more suitable to your particular case.
+>
+> Having said that, **enabling dependency injection in your code is a must to be able to use any of the test doubles**, including the mocks we describe next, so make sure you fully understand what it means and how write your code *the right way*.
+{: .callout}
+
+## Introducing Google Mock
+
+[Google Mock], or gMock, is a framework for creating mock **classes** and using them in C++. A mock object implements the same interface as a real object (so it can be used as one), but lets you specify at run time how it will be used and what it should do, setting expectations on these interactions.
```

Comment:
```suggestion
[Google Mock], or gMock, is a framework for creating mock **classes** and using them in C++. A mock class implements the same interface as the real class (so it can be used as one), but lets you specify how it will be used and what it should do at runtime, setting expectations on these interactions.
```